Detectors/EMCAL/reconstruction/src/Clusterizer.cxx

```diff
@@ -0,0 +1,180 @@
+// Copyright CERN and copyright holders of ALICE O2. This software is
+// distributed under the terms of the GNU General Public License v3 (GPL
+// Version 3), copied verbatim in the file "COPYING".
+//
+// See http://alice-o2.web.cern.ch/license for full licensing information.
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file Clusterizer.cxx
+/// \brief Implementation of the EMCAL clusterizer
+#include <cstring>
+#include "FairLogger.h" // for LOG
+#include "EMCALReconstruction/Clusterizer.h"
+
+using namespace o2::emcal;
+
+///
+/// Constructor
+//____________________________________________________________________________
+Clusterizer::Clusterizer(double timeCut, double timeMin, double timeMax, double gradientCut, bool doEnergyGradientCut, double thresholdSeedE, double thresholdCellE) : mSeedList(), mDigitMap(), mCellMask(), mTimeCut(timeCut), mTimeMin(timeMin), mTimeMax(timeMax), mGradientCut(gradientCut), mDoEnergyGradientCut(doEnergyGradientCut), mThresholdSeedEnergy(thresholdSeedE), mThresholdCellEnergy(thresholdCellE)
+{
+}
+
+///
+/// Recursively search for neighbours (EMCAL)
+//____________________________________________________________________________
+void Clusterizer::getClusterFromNeighbours(std::vector<Digit*>& clusterDigits, int row, int column)
+{
+  // Recursion 0, add seed digit to cluster
+  if (!clusterDigits.size())
+    clusterDigits.emplace_back(mDigitMap[row][column]);
+
+  // Mark the current cell as clustered
+  mCellMask[row][column] = kTRUE;
+
+  // Now go recursively to the next 4 neighbours and add them to the cluster if they fulfill the conditions
```

Comment:
instead all these if's, why don't you simply use
constexpr int rowDiff[4] = {-1,0,0,1}, colDiff[4] = {0,-1,1,0};
for (int dir=4;dir--;) {
  int rowD = row + rowDiff[dir], colD = column + collDiff[dir];
  // use rowD and colD instead of row + rowDiff, column + colDiff
}

Reply:
Done.